File: ios/Classes/model/map/overlay/overlay/LazyOrAddableOverlay.swift
internal protocol LazyOrAddableOverlay {
    associatedtype OverlayType: Any
    
    func createMapOverlay() -> OverlayType
}

internal func asLazyOverlayOrAddableOverlayFromMessageable(
    info: NOverlayInfo,
    args: Dictionary<String, Any>
) throws -> any LazyOrAddableOverlay {
    if (info.type.isLazy) {
        return try asLazyOverlayFromMessageable(info: info, args: args)
    }
    return try asAddableOverlayFromMessageable(info: info, json: args)
}

File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/LazyOrAddableOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import android.content.Context
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo

internal sealed interface LazyOrAddableOverlay {
    fun createMapOverlay(): Any

    companion object {
        fun fromMessageable(
            info: NOverlayInfo,
            args: Map<String, Any>,
            context: Context,
        ): LazyOrAddableOverlay {
            if (info.type.isLazy) return LazyOverlay.fromMessageable(info, args)
            return AddableOverlay.fromMessageable(info, args, context)
        }
    }
}
